Angular DI post (#1277)
oktadeveloper/okta-blog: _source/_posts/2022-10-11-angular-dependency-injection.md

---
layout: blog_post

> title: "Practical Usages of Dependency Injection in Angular"

Reviewer: I'm not going to edit the title, it's not wrong, and my comment is opinionated for sure, but I'm not a huge fan of "usage" - use is so much simpler/more direct, and since you're going for the plural, I'd recommend either "Practical Usage of Dependency Injection in Angular" (because usage can imply the plural without adding a syllable) OR "Practical Uses of Dependency Injection in Angular" - the scores are all identical

Author: I'm good with "uses" I'll make the change

When Angular CLI generates a service, it adds an `@Injectable()` TypeScript decorator, which is the bit of code that registers a service within Angular DI system:

```ts
@Injectable({
  providedIn: 'root'
})
export class MyService {
}
```

Without doing anything else, we can use our dependency in the application by injecting it into the consuming code as a constructor parameter:

```ts
@Component({
  // standard component metadata here
})
export class MyComponent {
  constructor(private myService: MyService) { }
}
```

> In Angular v14, you have a new option to use the `inject()` function instead of injecting the service into the consumer as a constructor parameter.

Another way to register dependencies is to provide them manually through the `providers` array. Different Angular building blocks accept `providers` in the metadata. So you can register a provider like this:

```ts
@NgModule({
  imports: // stuff here
  declarations: // stuff here
  providers: [
    MyService
  ]
})
export class AppModule {
}
```

> There's something else to note, though. Angular's DI system allows you to provide a dependency to different places within the application. We saw an example of this in the first code snippet of the `@Injectable()` TypeScript decorator. Angular CLI automatically generates

Reviewer: ```suggestion
There's something else to note, though. Angular's DI system allows you to provide a dependency to different places within the application. We saw an example of this in the first code snippet of the `@Injectable()` TypeScript decorator. Angular CLI automatically generates: 
```
adding a colon

```ts
@Injectable({
  providedIn: 'root'
})
export class MyService {
}
```

> The configuration option `providedIn: 'root'` specifies where within the application to provide the service. In this case, we're saying provide to "root", which means the root of the app so this service is available across the entirety of the application.

Reviewer: ```suggestion
The configuration option `providedIn: 'root'` specifies where within the application to provide the service. In this case, we're saying provide to "root", which means the root of the app, so this service is available across the entirety of the application.
```

 While having this level of configurability sounds unnecessarily complicated, it allows you to fine-tune which dependency to use in your consuming code. You can configure the `providers` array in other modules and Angular building blocks, such as components and directives. You'll get the provider you configured closest to the consuming code when you do so.

Now that we have a quick overview of how to provide dependencies, let's review an integral piece of Angular's DI system, injection tokens.

## Injection tokens in Angular

> Injection tokens allow us to have values and objects as dependencies. This means we can depend on a string, such as "Hello world!", and objects, which include configuration objects and global variables such as Web APIs. But injection tokens are even more remarkable because we can also create dependencies to constructs that don't have a runtime definition, such as interfaces! Let's take a look at an example using an injection token.

Reviewer: ```suggestion
Injection tokens allow us to have values and objects as dependencies. This means we can depend on strings, such as "Hello world!", and objects, which include configuration objects and global variables such as Web APIs. But injection tokens are even more remarkable because we can also create dependencies to constructs that don't have a runtime definition, such as interfaces! Let's take a look at an example using an injection token.
```
I think you're saying "we can depend on strings and objects" right? If so, it's clearer to the reader if both are plural... If I've misunderstood, I guess that proves my point ;-)

})
export class MyBreadComponent {
  constructor(@Inject(BREAD_TOKEN) private bread: string) { }
}
```

Now we can access the 🍞 emoji from within the component! This might not seem like a big deal, but having injection tokens as a means to represent values and interfaces as dependencies is enormous! And it sets us up to leverage the power of Angular's DI system.

We can use injection tokens along with configuring providers within Angular's DI system for more power and fine-grained control.

## Configuring providers in Angular's Dependency Injection system

You can configure the `providers` array to add fine-grained control to your providers. When combined with injection tokens, we can unleash a lot of power. But first, it's essential to know when it makes sense to do so. Always prefer the most straightforward, default way of registering a dependency and then use fine-grained control as needed.

To configure the `providers` array, you add an object containing the instructions like this:

```ts
@NgModule({
  imports: // stuff here
  declarations: // stuff here
  providers: [
    { provide: MyService, howToProvide: OtherDependency }
  ]
})
export class AppModule {
}
```

> The "how to provide" gives Angular-specific instructions on this dependency configuration. Then you can provide the other new dependency. Angular supports the following options for "how to provide":

Reviewer: Any reason here not to use code markup for "how to provide" ?

Author: i was keeping the code markup for real options so that it wouldn't be confusing, but if you think it's clearer to use code markup here I can

1. `useClass` - Replace the current dependency with a new instance of something else
2. `useExisting` - Replace the current dependency with an existing dependency
3. `useValue` - Replace the current dependency with a new value
4. `useFactory` - Use a factory method to determine which dependency to use based on a dynamic value

Next, let's walk through examples of each configuration option to understand how to use them.

### Configure providers with `useClass`

The `useClass` option replaces the current dependency with a new instance of another class. This is a great option to use if you're refactoring code and want to substitute a different dependency in your application quickly. Let's say you have a language learning app and an Angular service that wraps the authentication calls you delegate to an auth library and an auth provider. We'll call this service `AuthService`, and to keep things straightforward, the code looks like this:

```ts
@Injectable({
  providedIn: 'root'
})
export class AuthService {
  public login(): void { }
  public logout(): void { }
}
```

> In a stroke of luck, a large tech company decides to buy your language learning app, but they require you to authenticate using their social login only. You can create a new authentication service; we'll call it `NewAuthService` (note, you should not name your services with these terrible names), and keep the same member names that wraps the calls to their auth provider. 

Reviewer: ```suggestion
In a stroke of luck, a large tech company decides to buy your language learning app, but they require you to authenticate using their social login only. You can create a new authentication service that wraps the calls to their auth provider and keeps the same member names; we'll call it `NewAuthService`. (Note, you should not name your services with these terrible generic names. Be a bit more descriptive. )
```
I tried to simplify the last long sentence because I found it confusing. Hopefully I didn't misunderstand it... n

```ts
@Injectable({
  providedIn: 'root'
})
export class NewAuthService {
  public login(): void { /* new way to login */ }
  public logout(): void { /* new way to logout */ }
```

Because both classes have the same public members, you can substitute the original `AuthService` with the new `NewAuthService` by configuring the provider:

```ts
@NgModule({
  imports: // imports here
  declarations: //declarations here
  providers: [
    { provide: AuthService, useClass: NewAuthService }
  ]
})
export class AppModule { }
```

The cool thing about having the same public members is that there's no need to change the consuming code. Angular instantiates a new instance of `NewAuthService` and provides that dependency to consuming code, even if they still refer to `AuthService`! 

> It might not make sense to keep the original `AuthService` around, so eventually you might want to consider transferring all the code references to use the `NewAuthService` only. However, the `useClass` configuration option is a fast way for us to quickly substitute one instance of a class for another, which means proof-of-concepts and quick checks can be as fast as a drop of a hat!

Reviewer: ```suggestion
It might not make sense to keep the original `AuthService` around, so eventually you might want to consider transferring all the code references to use the `NewAuthService` only. However, the `useClass` configuration option is a fast way for us to quickly substitute one instance of a class for another, which means proof-of-concepts and quick checks can be super-fast!
```
I try to avoid idioms that might be difficult for non-native readers to understand.

@Component({
  // standard component metadata here
})
export class ElvishTranslationComponent implements OnInit {
  private elvish!: string;
  constructor(private translationService: LanguageTranslationService) { }
  
  public ngOnInit(): void {
    this.elvish = this.translationService.elvish(someText);
  }
}
```

Oops... The `LanguageTranslationService` looks a bit unwieldy. Let's narrow the API surface by creating a new class called `FictitiousLanguageTranslationService` and move the translation methods for the fictitious languages there. We'll use an abstract class for this:

```ts
export abstract class FictitiousLanguageTranslationService {
  abstract elvish: (text: string) => string;
  abstract klingon: (text: string) => string;
}
```

Now we can add `FictitiousLanguageTranslationService` as a real dependency in the application by adding it to the `providers` array, but use the existing `LanguageTranslationService` implementation of the code:

```ts
@NgModule({
  imports: // imports here
  declarations: // declarations here
  providers: [
    { provide: FictitiousLanguageTranslationService, useExisting: LanguageTranslationService }
  ]
})
export class AppModule { }
```

> Now we'll update the consumer to use the new dependency:

Reviewer: ```suggestion
Next we'll update the consumer to use the new dependency:
```
to avoid the repetition

In this example, let's say in your Language Learning application, if the user is learning French, we want to show a croissant in the `BreadComponent` instead of a loaf of bread. The user's language selection is in the user's config, so the example code looks like this:

```ts
@NgModule({
  imports: // imports here
  declarations: // declarations here
  providers: [
    {
      provide: BREAD_TOKEN,
      useFactory: (config: UserConfig) => config.language === 'fr' ? '🥐' : '🍞',
      deps: [UserConfig]
    }
  ]
})
export class AppModule { }
```

Notice we were able to pass in a dependency to the configuration option. The `useClass` and `useFactory` options support passing in dependencies.

Now when we use this in the `MyBreadComponent` we'll get a croissant instead of a loaf of bread only if the the user's configuration has French as their language!

```ts
@Component({
  // standard component metadata here
})
export class MyBreadComponent {
  constructor(@Inject(BREAD_TOKEN) private bread: string) { 
    // bread is either 🍞 or 🥐 based on the language setting
  }
}
```

## Dependency Injection when configuring external dependencies

> Now that we have a better understanding injection tokens and when to use the different configuration options, you may want to check out the post, [Three Ways to Configure Modules in Your Angular App](/blog/2022/02/24/angular-async-config). This post covers configuring modules and Okta configuration demonstrating the different concepts covered within in this post in real-life use cases.

Reviewer: ```suggestion
With a better understanding of injection tokens and when to use the different configuration options, next you may want to check out the post, [Three Ways to Configure Modules in Your Angular App](/blog/2022/02/24/angular-async-config). This covers configuring modules and Okta configuration demonstrating the different concepts covered within in this post in real-life use cases.
```
avoiding the "Now" opening...

## Learn more about Angular Dependency Injection

> This post is a high-level overview of Angular's DI system. As you can already see, this powerful system has many different configuration options and a lot of complexity. As a result, even though Angular has these configuration options, using the most straightforward approach will make troubleshooting and maintenance easier. 

Reviewer: ```suggestion
This article offers a high-level overview of Angular's DI system. As you can already see, it's a powerful system with many different configuration options and a lot of complexity. As a result, even though Angular has these configuration options, using the most straightforward approach will make troubleshooting and maintenance easier. 
```
